test/distributed/_spmd/test_data_parallel.py

```diff
@@ -0,0 +1,163 @@
+# Owner(s): ["oncall: distributed"]
+
+from copy import deepcopy
+
+import torch
+
+import torch.nn as nn
+
+from torch.distributed._spmd.api import compile
+
+from torch.distributed._spmd.parallel_mode import DataParallel
+from torch.distributed._tensor import Replicate
+from torch.nn.parallel import DistributedDataParallel as DDP
+from torch.testing._internal.common_distributed import skip_if_lt_x_gpu
+from torch.testing._internal.common_utils import run_tests
+from torch.testing._internal.distributed._tensor.common_dtensor import (
+    DTensorTestBase,
+    with_comms,
+)
+
+
+class SimpleMLP(nn.Module):
+    def __init__(self):
+        super(SimpleMLP, self).__init__()
+        self.net1 = nn.Linear(50, 32)
+        self.relu = nn.ReLU()
+        self.net2 = nn.Linear(32, 8)
+
+    def forward(self, x):
+        return torch.sigmoid(self.net2(self.relu(self.net1(x))))
+
+    def reset_parameters(self, *args, **kwargs):
+        self.net1.reset_parameters()
+        self.net2.reset_parameters()
+
+
+# simple train step definition, just an example
+def train_step(model, optim, train_batch):
+    def loss_fn(out, labels):
+        return (out - labels).sum()
+
+    optim.zero_grad()
+    inputs, labels = train_batch
+
+    out = model(inputs)
+    loss = loss_fn(out, labels)
+    loss.backward()
+    optim.step()
+    return loss
+
+
+class TestDataParallel(DTensorTestBase):
+    @property
+    def world_size(self):
+        return 2
+
+    def _test_data_parallel(
+        self, mod, ddp_mod, opt, ddp_opt, inp, train_step, data_parallel_mode
+    ):
+        ddp_inp = deepcopy(inp)
+
+        # need one step to warm up optimizers
+        train_step(mod, opt, inp)
+        opt.zero_grad()
+        # DDP run full train step once to align with the warmup
+        train_step(ddp_mod, ddp_opt, ddp_inp)
+        ddp_opt.zero_grad()
+
+        # train a DDP model once manually as DDP grads are different
+        torch.sum(ddp_mod(ddp_inp[0]) - ddp_inp[1]).backward()
+        with torch.no_grad():
+            for p in ddp_mod.parameters():
+                p.grad *= self.world_size
```

Comment:
Curious: Is there a reason why compile-mode data parallel does not average gradients by world size?

Reply:
I was following the other tests in `test_tracing.py` where it does not average gradients for the DTensorExpandMode, so honestly idk why we didn't do it, maybe @mrshenli have a better idea there. If there's no specific reason, I think we probably should average the gradients by default, I'll do it in a follow up PR.